### PR TITLE
feat(gateway): more explicit IPFSBackend and no multi-range

### PR DESCRIPTION
**Pull request by aschmahmann**
Also, switches to only handling single range requests rather than multiples.

 - [x] Merge conformance tests updates: https://github.com/ipfs/gateway-conformance/pull/113
 - [x] Release conformance tests updates
 - [x] Run new conformance tests
 - [x] Kubo PR for removing outdated sharness: https://github.com/ipfs/kubo/pull/9999
 - [x] Kubo PR https://github.com/ipfs/kubo/pull/10024

### CODE REVIEW

**Comment by hacdias** [comment]
@aschmahmann can we not support multiple ranges? Isn't that part of the specs? Also note that I'll be working on IPIP-412 soon, sow e might get some _conflicts_.

**Comment by BigLep** [comment]
2023-07-11 maintainer conversation: we shouldn't stress about multiple ranges.  We have a metric for this against ipfs.io and haven't found any actual usage.  This conformance test for multiple ranges can be moved to Kubo.
@aschmahmann will flip it out of draft when ready for review.

**Review by hacdias — CHANGES_REQUESTED** [review]
~Well, we still need the multiple range requests to pass conformance: https://github.com/ipfs/kubo/actions/runs/5655778072/job/15321538379?pr=10024~

Nevermind,  I just saw "This conformance test for multiple ranges can be moved to Kubo.".

**Comment by hacdias** [comment]
> This conformance test for multiple ranges can be moved to Kubo.

@BigLep @aschmahmann I don't understand how that helps. If we move the test to Kubo, it will still fail since we are removing the functionality on this PR and that functionality will simply no longer be available.

**Comment by aschmahmann** [comment]
> If we move the test to Kubo, it will still fail since we are removing the functionality on this PR and that functionality will simply no longer be available.

Correct. IIUC the right thing to do here is:
1. Have the conformance tests test the HTTP guarantees around multiple ranges (e.g. that the server can send back the ranges, the first range, or none as long as it's clear from the response headers)
2. If we end up having this functionality in just one implementation/consumer of boxo (e.g. kubo) then we test there

We can add back in the multiple range functionality, but IIUC from @lidel basically no one is using this so it's not worth the hassle at the moment.

**Comment by hacdias** [comment]
Working on updating the tests here: https://github.com/ipfs/gateway-conformance/pull/113

**Comment by hacdias** [comment]
@aschmahmann @lidel this PR is blocked until https://github.com/ipfs/gateway-conformance/pull/113 is merged and released (ready for review), so that we can have the tests passing here and/or in Kubo's PR.

File: gateway/blocks_backend.go
Comment (lidel) [inline gateway/blocks_backend.go]: @hacdias mind adding a PR (just PR, no need for IPIP since this is only a clarification) against https://specs.ipfs.tech/http-gateways/path-gateway/#range-request-header that adds one sentence that states  the implementation SHOULD support single ranges, and that support for more than one range is optional?

File: gateway/blocks_backend.go
Comment (lidel) [inline gateway/blocks_backend.go]: :thought_balloon: feels like `ImmutablePath` should have `func` that returns the rootCid, instead of parsing this everywhere like this

File: gateway/blocks_backend.go
Comment (lidel) [inline gateway/blocks_backend.go]: nit:  `[]cid.Cid{cid.MustParse("bafkqaaa")` could be a reusable const with comment linking to https://ipld.io/specs/transport/car/carv1/#number-of-roots

File: gateway/blocks_backend.go
Comment (lidel) [inline gateway/blocks_backend.go]: :thought_balloon:  TODO:  was not obvious to me how does 404 scenarios work.

When part of the requested content path does not exist, gateway should return CAR with  HTTP status code 200 and a CAR body that includes blocks up to the last available parent entity.

But here we return ReadCloser with `LastSegment: ifacepath.NewResolvedPath(ip, rootCid, rootCid, "")`.
What happens next?

File: gateway/errors.go
Comment (lidel) [inline gateway/errors.go]: :warning: I realized we have no end-to-end tests for 404 scenarios, just unit tests which do not really reflect final behavior.  Filled:

https://github.com/ipfs/gateway-conformance/issues/126 (car responses)
https://github.com/ipfs/gateway-conformance/issues/127 (deserialized behavior)


I feel we need to add them as part of this work. @hacdias would you have bandwidth to add them?

File: gateway/handler_codec.go
Comment (lidel) [inline gateway/handler_codec.go]: nit: move to `serve_http_content.go`

**Review by lidel — CHANGES_REQUESTED** [review]
(submitting first-pass at the end of day)

File: gateway/blocks_backend.go
Comment (hacdias) [reply gateway/blocks_backend.go]: https://github.com/ipfs/specs/pull/430

File: gateway/blocks_backend.go
Comment (hacdias) [reply gateway/blocks_backend.go]: I couldn't find many other similar cases like this. I think this is something we can solve with #334.

File: gateway/blocks_backend.go
Comment (hacdias) [reply gateway/blocks_backend.go]: @lidel as far as I understand, this will create a CAR file with all the blocks to resolve the path up to the last available entity.

File: gateway/errors.go
Comment (hacdias) [reply gateway/errors.go]: @lidel not today. Maybe Friday or early next week.

Comment (hacdias) [reply gateway/errors.go]: @lidel 
- https://github.com/ipfs/gateway-conformance/pull/131
- https://github.com/ipfs/gateway-conformance/pull/132

**Review by hacdias — APPROVED** [review]
Tests are passing in https://github.com/ipfs/kubo/pull/10024.

File: gateway/blocks_backend.go
Comment (lidel) [inline gateway/blocks_backend.go]: @hacdias @aschmahmann  Could we keep a meaningful root for exports done with `/ipfs/cid` (without any sub-path)?  

This would be a more gentle way of doing this, as we would not break existing users (pre IPIP-402) who rely on exporting full DAG by its CID, and  then doing `ipfs dag import` without extra pinning to keep it around (because `ipfs dag import` pins roots by default).

Comment (aschmahmann) [reply gateway/blocks_backend.go]: Yeah, that seems fine by me.

Comment (hacdias) [reply gateway/blocks_backend.go]: Hmm. The line you highlighted is only for paths that cannot be resolved - the other CARs (line 344) use the CID of the last segment as the root CID. Only if `/ipfs/{cid}` fails to resolve with a `isErrNotFound` is this code path followed. Is that a realistic scenario?

File: gateway/gateway.go
Comment (lidel) [inline gateway/gateway.go]: nit:  `HeadResponse` field introduced in this PR should match already existing  `GetResponse.byteSize` 
```suggestion
	bytesSize          int64
```

File: gateway/handler_defaults.go
Comment (lidel) [inline gateway/handler_defaults.go]: (example how it looks weird, new type should use the same field name)

File: gateway/serve_http_content.go
Comment (lidel) [inline gateway/serve_http_content.go]:
```suggestion
// parseRange parses a Range header string as per RFC 7233.
// errNoOverlap is returned if none of the ranges overlap.
func parseRange(s string, size int64) ([]httpRange, error) {
```

File: gateway/serve_http_content.go
Comment (lidel) [inline gateway/serve_http_content.go:1]: :thought_balloon:  / :warning: something that makes me anxious is that we have no(?) tests for all the low level HTTP funcs related to `httpServeContent` that we've ended up vendoring here. 

In the past we had the protection of unit tests from places like https://cs.opensource.google/go/go/+/master:src/net/http/range_test.go  https://cs.opensource.google/go/go/+/master:src/net/http/fs_test.go (and others in the dir), but now we don't use the same funcs as upstream golang and lose some coverage.

For example (full list would require bit closer look), conformance tests exist for some  If-None-Match scenarios, but we have no tests for behaviors around  If-Match, If-Unmodified-Since, If-Modified-Since, and If-Range.

We could add basic unit tests here, or invest time into better coverage in gateway conformance to benefit implementations other than  boxo/gateway (fill issue for missing tests in conformance repo).

@aschmahmann any preference?

File: gateway/handler_unixfs_dir.go
Comment (aschmahmann) [inline gateway/handler_unixfs_dir.go]: @lidel @hacdias how's this for our `//index.html` issues? Kind of gross, but seems like it'd do the job until we have something like #334 to make this less painful.

File: gateway/blocks_backend.go
Comment (aschmahmann) [reply gateway/blocks_backend.go]: Good call, I wasn't looking too closely when I commented here. We shouldn't be able to hit this codepath with a failed `/ipfs/cid` resolution, and if we did we should be returning an error not writing out an empty CAR file.

**Review by lidel — CHANGES_REQUESTED** [review]

File: gateway/handler_unixfs_dir.go
Comment (lidel) [reply gateway/handler_unixfs_dir.go]: sgtm. we could would flip it to append index.html only in one place:

```suggestion
	appendIndexHtml := func(p ipath.Path) ipath.Path {
		basePath := p.String()
		if basePath[len(basePath)-1] != '/' {
			basePath += "/"
		}
		return ipath.New(basePath + "index.html")
	}

	idxPath := appendIndexHtml(contentPath)
	imIndexPath, err := NewImmutablePath(appendIndexHtml(resolvedPath))
```

File: gateway/handler_codec.go
Comment (aschmahmann) [inline gateway/handler_codec.go]: @hacdias I don't know if you or I put this in, but this looks  suspicious. Why would we be doing manipulation of the data range before doing the full rendering/transformation?

Comment (hacdias) [reply gateway/handler_codec.go]: The blame says you added it - tbf, I do not remember adding it, so it's probably right. (698eccfdefb70d57fc5cab8cbf36f47afb00b25b). I have no idea about what it might be doing.

**Comment by BigLep** [comment]
2023-08-22 conversation: @lidel will review and figure out appropriate testing story (Boxo unit tests vs. conformance tests)

**Comment by BigLep** [comment]
2023-08-24 conversation: 
We are going to lean on conformance tests.  @lidel is going to create a gateway-conformance issues for this.
It's a lesser evil to merge without the conformance tests than keep this open and let it diverge.
@aschmahmann will add some Boxo unit tests and flag the changed lines of potential concerns.
@aschmahmann will make sure we get the merge decision 
(There is also some dance for keeping CI green with sharness.)

**Comment by BigLep** [comment]
@aschmahmann : is this on track to land this week?

**Comment by BigLep** [comment]
2023-09-12 conversation:
- Do a gateway-conformance release as soon as possible: https://github.com/ipfs/gateway-conformance/pull/162 lands and passes.
- To get green CI
- [x]  Conformance release
- [x]  Bubble up to Boxo
- [x]  Bubble this PR into Kubo
- [x]  Run this PR against Kubo version
- [x]  Once Boxo CI passes
- [x]  Switch to target Kubo master
- [x]  Merge this PR
- [x]  Bubble up Boxo commit to Kubo PR
- [x]  Merge Kubo PR
- [x]  Rerun Boxo CI

**Comment by hacdias** [comment]
CI is green in Kubo. As per what was agreed, I'm merging this once CI finishes here and bubbling it up.

**Pull request merged**
